Store Start Pages independently of Flow content

chrisroos opened the pull request:
This is an alternative approach to PR #3122.

This PR splits each Smart Answer into a Start Page and a Flow in terms of the Publishing API/Content Store. It publishes the content of the Start Page as a `transaction` and defines the `rendering_app` to be `frontend`. I believe this achieves the main goal of being able to render the Smart Answer Start Pages (along with most/all other Start Pages) through the Frontend app.

In contrast to PR #3122, this branch continues to set the `publishing_app` of the Start Pages as `smartanswers`. This avoids us having to forcefully reserve routes in order to change the `publishing_app` to `Publisher` each time we deploy Smart Answers. That change can happen separately and should probably be a one-off task that's executed within the Publisher environment (something like get Start Page content from Publishing API, add corresponding Transaction artefact to Publisher, reserve the route for Publisher, publish the content through Publisher).

I've deployed this branch to Integration and visually checked each of the Smart Answers using the [links in this gist](https://gist.github.com/chrisroos/a9a0fb6c4c27cf7b472464edf349d66b). Aside from the cosmetic change to the "Start now" button, everything else appears to be identical. The only difference is with the [maternity-paternity-calculator](https://www.gov.uk/maternity-paternity-calculator). The maternity-paternity-calculator start page now includes "Before you start" below the start button instead of "What you need".

I've highlighted the differences mentioned above in the screenshots below.

## maternity-paternity-calculator

![smart-answers-maternity-paternity-calculator-start-page](https://user-images.githubusercontent.com/6556/27869003-754cca9c-6196-11e7-8d13-bdd8209afe94.png)

![screenshot 2017-07-06 10 53 29](https://user-images.githubusercontent.com/6556/27905969-98b0a7da-6239-11e7-974b-5dcb7aad2bb5.png)

chrisroos commented:
Thanks @h-lame :-)

chrisroos commented:
We've had confirmation from the content team (via Humin) that the "What you need" text below the start button on the maternity-paternity-calculator can be removed. I've added a commit to remove this text and updated the screenshot in the PR description to show the difference.

binaryberry commented on app/presenters/start_page_content_item.rb:
```diff
@@ -0,0 +1,61 @@
+            content: flow_presenter.start_page_post_body,
+            content_type: 'text/govspeak'
+          ],
+          transaction_start_link: base_path + '/y'
```
Could we add the start_button_text field as part of this PR? My understanding is now that not being able to transfer the customised button text for child-maintenance would stop us from deploying, so it would make sense to include in this PR?

pmanrubia commented:
This is great work 👍 
I love the idea of splitting into two steps the process of using `transaction` documents for start pages. It seems that once we are able to set the button text to its original value for `Calculate your child maintenance` we should be good to merge.

chrisroos commented:
I've opened PR #3129 which makes the custom start button text available to the `StartNodePresenter`. We'll need to merge those changes before we can update this PR to start sending the custom start button content to the Publishing API.

@pmanrubia: Are you in a position to review PR #3129? Or is there someone else better placed to do it?

chrisroos commented:
I've updated this PR to make use of the changes in PR #3129. This branch also now sends the start button text to the Publishing API. NOTE that the tests will fail until the changes in https://github.com/alphagov/govuk-content-schemas/pull/633 (adding support for start_button_text) are merged.

ikennaokpala commented:
@chrisroos I observed a few cosmetic changes that could help the narrative of some the tests a bit better in `test/unit/start_page_content_item_test.rb` lines 63 and 104. Also IMHO this PR presented the opportunity to do this "#{/flow_presenter.slug/y}" on line 38 `app/presenters/flow_content_item.rb`

binaryberry commented:
@chrisroos I've now deployed the govuk-content-schemas changes and re-ran the CI tests on this PR and they pass. This branch does need a rebase though.

chrisroos commented:
Thanks, @binaryberry. I'm going to rebase this on master and deploy to integration shortly to check that it all works as expected.

chrisroos commented:
Thanks for the feedback, @ikennaokpala. I've addressed your points below:

> I observed a few cosmetic changes that could help the narrative of some the tests a bit better in test/unit/start_page_content_item_test.rb lines 63 and 104 #payload...

Good spot! I've updated the test names so that they consistently use `#payload`.

> Also IMHO this PR presents the opportunity to do this "/#{flow_presenter.slug}/y" on line 38 app/presenters/flow_content_item.rb and line 52 on app/presenters/start_page_content_item.rb

I don't feel particularly strongly about this change so have decided to leave it as-is to reduce the number of changes in this branch.

> and line 5 on lib/smart_answer_flows/additional-commodity-code.rb consider single vs double quotes.

I have a personal preference for single quotes where possible but, as above, have decided to leave them to reduce the number of changes in the branch.

ikennaokpala commented:
@chrisroos 👍  Those were nice to haves.. The overall implementation is beauty to the eyes of the beholder.

chrisroos commented:
We've just deployed this branch to integration and confirmed that the custom start button text for calculate-your-child-maintenance appears as expected:

![screenshot 2017-07-11 13 05 49](https://user-images.githubusercontent.com/6556/28067565-b75ababe-6639-11e7-95a4-960de5de8ccb.png)

chrisroos commented:
@nickcolley received confirmation from Humin that this was good to merge.
